### PR TITLE
Monte Carlo tree search for RL

### PR DESCRIPTION
**Pull request by peastman**
Implements #890.  I'm still cleaning up some aspects, but if anyone wants to try this out, I'd really appreciate your feedback.  How well does it work for you?

### CODE REVIEW

File: deepchem/rl/mcts.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+      if not self._state_is_list:
+        state = [state]
+      buffer.append((state, prob, reward))
+      action = np.random.choice(np.arange(self._env.n_actions), p=prob)
+      self._env.step(action)
+      if self._env.terminated:
+        self._env.reset()
+        root = TreeSearchNode(0.0)
+      else:
+#        root = root.children[action]
```
Comment (proteneer) [inline deepchem/rl/mcts.py]: could probably move root = TreeSearchNode(0.0) to outside of if block

File: deepchem/rl/mcts.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+  def _create_trace(self, env, node, trace):
+    """Create one trace as part of the tree search."""
+    trace.append(node)
+    node.count += 1
+    if env.terminated:
+      # Mark this node as terminal
+      node.children = None
+      node.value = 0.0
+      return
+    if node.children is not None and len(node.children) == 0:
```
Comment (proteneer) [inline deepchem/rl/mcts.py]: Under what circumstances is children None and this line is still reachable? The check for if node.children is not None seems unnecessary since once node.children is set to None then it returns.

File: deepchem/rl/mcts.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+      # Expand this node.
+      prob_pred, value = self.predict(env.state)
+      node.value = float(value)
+      node.children = [TreeSearchNode(p) for p in prob_pred[0]]
+    if len(trace) == self.max_search_depth:
```
Comment (proteneer) [inline deepchem/rl/mcts.py]: To be clear, does this generate self.max_search_depth + 1 layers in search tree? I was wondering why you didn't check this first before expanding.

File: deepchem/rl/mcts.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+            self._graph.session, self._graph.save_file, global_step=self._checkpoint_index)
+        self._checkpoint_time = time.time()
+    return buffer
+
+  def _optimize_policy(self, buffer, epochs):
+    """Optimize the policy based on the replay buffer from the current iteration."""
+    batch_size = self._graph.batch_size
+    n_batches = len(buffer) // batch_size
+    for epoch in range(epochs):
+      np.random.shuffle(buffer)
```
Comment (proteneer) [inline deepchem/rl/mcts.py]: What's the rationale for the shuffle? Note that this mutates a passed in argument (Though here it doesn't matter because the caller doesn't use buffer anymore).

File: deepchem/rl/mcts.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+          feed_dict = {}
+          for i, f in enumerate(self._features):
+            feed_dict[f] = np.stack(buffer[j][0][i] for j in indices)
+          feed_dict[self._search_prob] = np.stack(buffer[j][1] for j in indices)
+          feed_dict[self._search_value] = np.array([buffer[j][2] for j in indices])
+          yield feed_dict
+
+      loss = self._graph.fit_generator(generate_batches(), checkpoint_interval=0)
+
+  def _do_tree_search(self, root, temperature, adapt_puct):
```
Comment (proteneer) [inline deepchem/rl/mcts.py]: I'm trying to understand this algorithm a little bit better.

It looks like it basically does an alternating BFS/DFS, always starting from the root. The rewards are only recorded along the "trace", i.e. the chosen path down this tree. All other off-trace nodes have a reward of zero and do not contribute towards the root's count/rewards. Am I getting this correct?

File: deepchem/rl/mcts.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+  def create_tensor(self, **kwargs):
+    pred_prob, pred_value, search_prob, search_value = [
+        layer.out_tensor for layer in self.in_layers
+    ]
+    log_prob = tf.log(pred_prob + np.finfo(np.float32).eps)
+    probability_loss = -tf.reduce_mean(search_prob * log_prob)
+    value_loss = tf.reduce_mean(tf.square(pred_value - search_value))
+    self.out_tensor = probability_loss + self.value_weight * value_loss
+    self.probability_loss = probability_loss
+    self.value_loss = value_loss
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/rl/mcts.py]: Debugging. :)  It's so I can query them separately if I want to.

File: deepchem/rl/mcts.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+  of two stages:
+
+  1. Simulate many episodes.  At every step perform a tree search to determine
+     targets for the probabilities and value function, and store them into a
+     buffer.
+  2. Optimize the policy using batches drawn from the buffer generated in step 1.
+
+  The tree search involves repeatedly selecting actions starting from the
+  current state.  This is done by using deepcopy() to clone the environment.  It
+  is essential that this produce a deterministic sequence of states: performing
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/rl/mcts.py]: We're building a tree where each node corresponds to a state, and each child indicates what state you get to by selecting a particular action.  If that's not deterministic (at least within this one tree search), the whole algorithm falls apart.

File: deepchem/rl/mcts.py [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+               model_dir=None):
+    """Create an object for optimizing a policy.
+
+    Parameters
+    ----------
+    env: Environment
+      the Environment to interact with
+    policy: Policy
+      the Policy to optimize.  Its create_layers() method must return a map containing the
+      keys 'action_prob' and 'value', corresponding to the action probabilities and value estimate
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/rl/mcts.py]: No, just one estimate.

File: deepchem/rl/mcts.py (hunk 1 above)
Comment (peastman) [reply deepchem/rl/mcts.py]: As you can see by the commented out line, I'm still working on this bit of code.  I want to reuse results from the previous tree search, but I haven't gotten it working yet, so for the moment I just throw them out and create a new node.

File: deepchem/rl/mcts.py (hunk 4 above)
Comment (peastman) [reply deepchem/rl/mcts.py]: The same reason you would do a shuffle on any optimization: avoid artifacts by making the division into batches different for each epoch.

File: deepchem/rl/mcts.py (hunk 5 above)
Comment (peastman) [reply deepchem/rl/mcts.py]: The term "tree search" is misleading because it *isn't* really a tree search.  It runs a set of simple rollouts and builds a tree of the states that happened to be encountered by them, but it's not doing any sort of exhaustive exploration of all branches.

File: deepchem/rl/mcts.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (proteneer) [reply deepchem/rl/mcts.py]: Is it not possible to hash the state / actions somehow as ids of the node in the tree? It's a little hard to follow which of the random.shuffle / random.choice calls explicitly depend on this.

File: deepchem/rl/mcts.py (hunk 2 above)
Comment (peastman) [reply deepchem/rl/mcts.py]: More just a sanity check.  I set `children` to `None` to mark it as a terminal node.  Assuming the state transitions really are deterministic, if `env.terminated` was `True` the last time it got to this node, it should be again.  But just in case, I include that check to avoid an exception.

File: deepchem/rl/mcts.py (hunk 3 above)
Comment (peastman) [reply deepchem/rl/mcts.py]: If we've reached the maximum search depth, we need to set `node.value` as done just above.  Since we're not going to search any further, we fill in our best estimate of the rewards we *would* have gotten by continuing.

File: deepchem/rl/tests/test_mcts.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        self._state = [np.array([0])]
+
+      def step(self, action):
+        if action == 37:
+          self._terminated = True  # Walk away.
+          return 0.0
+        wheel = np.random.randint(37)
+        if wheel == 0:
+          if action == 0:
+            return 35.0
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/rl/tests/test_mcts.py]: Whoever wrote that was mistaken.  Here's the implementation of it: https://github.com/openai/gym/blob/master/gym/envs/toy_text/roulette.py

File: deepchem/rl/mcts.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/rl/mcts.py]: A state is defined by a set of (possibly very large) numpy arrays.  Trying to hash them isn't really practical.

File: deepchem/rl/tests/test_mcts.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (proteneer) [reply deepchem/rl/tests/test_mcts.py]: Thanks for the link, though that implementation is super confusing. It say it returns 35, but it actually returns 36. Since in the constructor:
```
def __init__(self, spots=37):
  self.n = spots + 1
```
then in _step:
```
if val == action == 0:
  reward = self.n - 2.0
```

File: deepchem/rl/mcts.py [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+               optimizer=Adam(),
+               model_dir=None):
+    """Create an object for optimizing a policy.
+
+    Parameters
+    ----------
+    env: Environment
+      the Environment to interact with
+    policy: Policy
+      the Policy to optimize.  Its create_layers() method must return a map containing the
```
Comment (lilleswing) [inline deepchem/rl/mcts.py]: knit: 
```
return a dict containing the
```

**Comment by proteneer** [comment]
What's the treatment for cycles in the state? For example given a game like Chess, a rook can move from i->j->i again. Currently it looks like would mark i->j->i' such that i and i' are different states despite an isomorphism.

**Comment by peastman** [comment]
It just unrolls it and ignores the fact that those states are identical.  I expect that situation to be pretty rare in practice, and ignoring it isn't wrong, it just means you miss a chance to get better statistics by combining them.

**Comment by peastman** [comment]
Something is messed up with yapf.  It's failing on the Python 2.7 build, despite that fact that I *did* run yapf locally before checking the code in.  Furthermore, the Python 3.5 build passes!

**Comment by peastman** [comment]
Aside from the problems with yapf, I think this is ready to merge.  I'll probably have more changes in the future, but right now I'd really like for people to start trying it out and giving feedback on how it works.

**Comment by rbharath** [comment]
Looked through and it is some yapf weirdness. Ideally, we should find a fix for this, but good to merge in this PR.

LGTM

**Comment by lilleswing** [comment]
everytime we have yapf weirdness its with the *args syntax.  Not that I have an answer just noting.

**Comment by peastman** [comment]
We could try upgrading to a newer version.  The latest version is 0.19, and we're still using 0.17.
